src/librustc_data_structures/sync.rs

```diff
@@ -241,6 +242,75 @@ impl<K: Eq + Hash, V: Eq, S: BuildHasher> HashMapExt<K, V> for HashMap<K, V, S>
     }
 }
 
+/// A type whose inner value can be written once and then will stay read-only
+pub struct Once<T>(Lock<Option<T>>, PhantomData<T>);
```

Comment:
What's the `PhantomData` for?

Reply:
Once conceptually owns a `T` outside the lock once it is initialized. This ensures we get the right impl for `Sync`: `impl<T: Sync> Sync for Once<T>`

Reply:
Good catch! Could you add a comment explaining that?

Reply:
Plus maybe and example of how one could exploit `Once<T>` if the `PhantomData` was not there?

Reply:
Done.